lightnode/kvstore/storeutil/ttl_wrapper.go

```diff
@@ -0,0 +1,272 @@
+// special expiry keys are stored with a prefix "e". The expiry key also contains the expiry time in hexadecimal format,
+// such that when iterating over expiry keys in lexicographical order, the keys are ordered by expiry time. The value
+// each expiry key points to is the regular key that is to be deleted when the expiry time is reached. In order to
+// efficiently delete expired keys, the expiry keys must be iterated over periodically to find and delete expired keys.
+type ttlStore struct {
+	store  kvstore.Store
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	logger logging.Logger
+}
+
+// TTLWrapper extends the given store with TTL capabilities. Periodically checks for expired keys and deletes them
+// with a period of gcPeriod. If gcPeriod is 0, no background goroutine is spawned to check for expired keys.
+func TTLWrapper(
+	ctx context.Context,
+	logger logging.Logger,
+	store kvstore.Store,
+	gcPeriod time.Duration) kvstore.TTLStore {
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	ttlStore := &ttlStore{
+		store:  store,
+		ctx:    ctx,
+		cancel: cancel,
+		logger: logger,
+	}
+	if gcPeriod > 0 {
+		ttlStore.expireKeysInBackground(gcPeriod)
+	}
+	return ttlStore
+}
+
+var keyPrefix = []byte("k")
+var expiryPrefix = []byte("e")
+var maxDeletionBatchSize = 1000
+
+// PutWithTTL adds a key-value pair to the store that expires after a specified time-to-live (TTL).
+// Key is eventually deleted after the TTL elapses.
+func (store *ttlStore) PutWithTTL(key []byte, value []byte, ttl time.Duration) error {
+	expiryTime := time.Now().Add(ttl)
+	return store.PutWithExpiration(key, value, expiryTime)
+}
+
+// buildExpiryKey creates an expiry key from the given expiry time.
+// The expiry key is composed of the following 3 components appended one after the other:
+// - a one byte "e" prefix
+// - the expiry time in hexadecimal format (8 bytes)
+// - and the base key.
+func buildExpiryKey(
+	baseKey []byte,
+	expiryTime time.Time) []byte {
+
+	expiryKeyLength := 1 /* prefix */ + 8 /* expiry timestamp */ + len(baseKey)
+	expiryKey := make([]byte, expiryKeyLength)
+
+	expiryKey[0] = 'e'
+	expiryUnixNano := expiryTime.UnixNano()
+	for i := 0; i < 8; i++ {
+		expiryKey[1+i] = byte((expiryUnixNano >> (56 - i*8)) & 0xff)
+	}
+
+	copy(expiryKey[9:], baseKey)
+
+	return expiryKey
+}
+
+// parseExpiryKey extracts the expiry time and base key from the given expiry key.
+func parseExpiryKey(expiryKey []byte) (baseKey []byte, expiryTime time.Time) {
+	expiryUnixNano := int64(0)
+	for i := 0; i < 8; i++ {
+		expiryUnixNano |= int64(expiryKey[1+i]) << (56 - i*8)
+	}
+	expiryTime = time.Unix(0, expiryUnixNano)
+
+	baseKey = expiryKey[9:]
+	return
+}
+
+// PutWithExpiration adds a key-value pair to the store that expires at a specified time.
+// Key is eventually deleted after the expiry time.
+func (store *ttlStore) PutWithExpiration(key []byte, value []byte, expiryTime time.Time) error {
```

Comment:
The client may want to expire the entire batch that was written atomically, as that may often be a logically meaningful group of KVs.

Reply:
I agree that this is a useful general purpose feature. I didn't add something like this because light nodes won't need it. Do you think I should add this feature as a part of this PR?

Reply:
If we want to unify different KV store wrappers, it may be needed. In the DA Node, it handles all blobs in a batch atomically (store and delete)

Reply:
I've implemented the requested `PutBatchWithExpiration()` method in anticipation of the future merge of the two implementations, but haven't actually merged the implementations as a part of this PR.